pkg/proxy/handlers/spacelister.go

```diff
@@ -34,25 +35,59 @@ func NewSpaceLister(app application.Application) *SpaceLister {
 }
 
 func (s *SpaceLister) HandleSpaceListRequest(ctx echo.Context) error {
-	userID, _ := ctx.Get(context.SubKey).(string)
-	username, _ := ctx.Get(context.UsernameKey).(string)
+
+	workspaces, err := s.ListUserWorkspaces(ctx)
+	if err != nil {
+		return errorResponse(ctx, apierrors.NewInternalError(err))
+	}
+
 	workspaceName := ctx.Param("workspace")
 	doGetWorkspace := len(workspaceName) > 0
 
+	if doGetWorkspace { // specific workspace requested
+		if len(workspaces) != 1 {
+			// not found
+			r := schema.GroupResource{Group: "toolchain.dev.openshift.com", Resource: "workspaces"}
+			return errorResponse(ctx, apierrors.NewNotFound(r, workspaceName))
+		}
+		return getWorkspaceResponse(ctx, workspaces[0])
+	}
+
+	return listWorkspaceResponse(ctx, workspaces)
```

Comment:
just cosmetics - using the `switch`  statement could be another option, something similar to:

```
    switch {
    case len(workspaces) != 1:
        // invalid request with multiple workspaces
        r := schema.GroupResource{Group: "toolchain.dev.openshift.com", Resource: "workspaces"}
		return errorResponse(ctx, apierrors.NewNotFound(r, workspaceName))
    case len(workspaces) == 1:
        // valid request for single workspace
        return getWorkspaceResponse(ctx, workspaces[0])
    default:
        // valid request to list all available workspaces 
        return listWorkspaceResponse(ctx, workspaces)
    }
```
just an alternative idea.

Reply:
Ah I see now, my bad! I got confused between `len(workspaceName)` and `len(workspaces)`. So basically if a workspace was requested but you have more then one in the list or zero workspaces you return a `not found` error. 

If not specific workspace was asked you provide the list , got it! 
Maybe if there is more then 1 workspaces in the list, the error message could be different one, not sure.